conctr.agent.lib.nut

```diff
@@ -0,0 +1,556 @@
+// MIT License
+
+// Copyright (c) 2016-2017 Mystic Pants Pty Ltd
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+class Conctr {
+
+    static VERSION = "2.0.0";
+
+    static DATA_EVENT = "conctr_data";
+    static LOCATION_REQ_EVENT = "conctr_get_location";
+    static AGENT_OPTS_EVENT = "conctr_agent_options";
+    static SOURCE_DEVICE = "impdevice";
+    static SOURCE_AGENT = "impagent";
+    static MIN_TIME = 946684801; // Epoch timestamp for 00:01 AM 01/01/2000 (used for timestamp sanity check)
+    static DEFAULT_LOC_INTERVAL = 3600; // One hour in seconds   
+    static MIN_RECONNECT_TIME = 5;
+
+
+    // Request opt defaults
+    static MAX_RETIES_DEFAULT = 3;
+    static RETRIES_LIMIT = 10;
+    static EXP_BACKOFF_DEFAULT = true;
+    static RETRY_INTERVAL_DEFAULT = 5;
+    static MAX_RETRY_INTERVAL = 60;
+
+    // Default location recording opts
+    static DEFAULT_LOC_ENABLED = false;
+    static DEFAULT_LOC_INTERVAL = 3600;
+    static DEFAULT_LOC_SEND_ONCE = true;
+    static DEFAULT_WAKE_REASONS = [];
+
+
+    // Location recording parameters
+    _locEnabled = null;     // Boolean to enable/disable location sending
+    _locInterval = null;    // Integer time interval between location updates
+    _locSendOnce = null;    // Boolean to send location only once
+    _locWakeReasons = null; // Array of hardware.wakereasons()
+
+
+    // Conctr Variables
+    _api_key = null;
```

Comment:
// Conctr Variables,  // Pending queue status, // Location state should have a single line comment